catalyst/dl/callbacks/criterion.py

```diff
@@ -1,44 +1,74 @@
+
+
+def _add_loss_to_state(
+    loss_key: Optional[str],
+    state: RunnerState,
+    loss: torch.Tensor
+):
+    if loss_key is None:
+        if state.loss is not None:
+            if isinstance(state.loss, list):
+                state.loss.append(loss)
+            else:
+                state.loss = [state.loss, loss]
+        else:
+            state.loss = loss
+    else:
+        if state.loss is not None:
+            assert isinstance(state.loss, dict)
+            state.loss[loss_key] = loss
+        else:
+            state.loss = {loss_key: loss}
+
 
 class CriterionCallback(Callback):
+    """
+    Callback for that measures loss with specified criterion.
+    """
     def __init__(
         self,
         input_key: str = "targets",
         output_key: str = "logits",
         prefix: str = "loss",
         criterion_key: str = None,
-        loss_key: str = None,
         multiplier: float = 1.0
     ):
+        """
+        Args:
+            input_key (str): key that takes values from the input dictionary
+                If None, the whole input will be passed to the criterion.
+            output_key (str): key that takes values from the output dictionary
+                If None, the whole output will be passed to the criterion.
+            prefix (str): prefix for metrics and output key for loss
+                in ``state.loss`` dictionary
+            criterion_key (str): A key to take a criterion in case
+                there are several of them and they are in a dictionary format.
+            multiplier (float): scale factor for the output loss.
+        """
         super().__init__(CallbackOrder.Criterion)
         self.input_key = input_key
         self.output_key = output_key
         self.prefix = prefix
         self.criterion_key = criterion_key
-        self.loss_key = loss_key
         self.multiplier = multiplier
 
-    def _add_loss_to_state(self, state: RunnerState, loss):
-        if self.loss_key is None:
-            if state.loss is not None:
-                if isinstance(state.loss, list):
-                    state.loss.append(loss)
-                else:
-                    state.loss = [state.loss, loss]
-            else:
-                state.loss = loss
-        else:
-            if state.loss is not None:
-                assert isinstance(state.loss, dict)
-                state.loss[self.loss_key] = loss
-            else:
-                state.loss = {self.loss_key: loss}
+    @staticmethod
+    def _get(dictionary: dict, key: Optional[str]) -> Any:
+        result = dictionary[key] if key is not None else dictionary
+        return result
 
     def _compute_loss(self, state: RunnerState, criterion):
-        loss = criterion(
-            state.output[self.output_key], state.input[self.input_key]
-        )
+        output = self._get(state.output, self.output_key)
+        input = self._get(state.input, self.input_key)
+
+        loss = criterion(output, input)
```

Comment:
as I remember you required to add KV-support. From my perspective, something like:
```python
def _get(dictionary: dict, key: Optional[str]) -> Any:
    result = {key_:dictionary[key_] for key_ in key} if key is not None else dictionary
    return result

output = self._get(state.output, self.output_key)
input = self._get(state.input, self.input_key)

loss = criterion(**output, **input)
```
should work

Reply:
u mean, add support for multiple `self.output_key` and `self.input_key`?

Reply:
yes, as u say "in detection, the loss may be counted with several keys."

Reply:
done !